docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+  to retain the same attachments as it had before restart, without having to map them to a different node ID.
+- We may in future run the pageserver in a different cluster runtime environment (e.g. k8s) where
+  guaranteeing that the cluster manager (e.g. k8s) doesn't run two concurrent instances with the same
+  node ID is much harder.
+
+#### Distinction between attachment and node generation id
+
+The most important generation number is the tenant attachment number: this alone would be sufficient
+to implement safe tenant migration and instance failover, if we assume that our control plane will never concurrently
+run two nodes with the same node ID.
+
+Running two nodes with the same ID might sound far-fetched, but it would happen very easily if we
+ran the pageserver in a k8s environment with a StatefulSet, as k8s provides no guarantee that
+an old pod is dead before starting a new one.
+
+The node generation is useful other ways, beyond it's core correctness purpose:
+
+- Including both generations (and the node ID) in object keys (see [object keys](#object-key-changes)) also provides some flexibility in
+  how HA should work in future: we could move to a model where failover can happen within one
+  attachment generation number (i.e. without control plane coordination) without changing the
+  storage format, because node IDs/generations would de-conflict writes from peers.
+- The node generation is also useful for managing remote objects which are not per-tenant,
+  such as the persistent per-node deletion queue which is proposed in this RFC.
+- node generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+Note that the two generation numbers have a different behavior for stale generations:
+
+- A pageserver with a stale generation number should immediately terminate: it is never intended
+  to have two pageservers with the same node ID. This is not necessary for correctness, as the
+  node generation appears in object keys, but objects written in a stale node generation are never
+  read later, so waste storage.
+- An attachment with a stale generation number is permitted to continue operating (ingesting WAL
+  and serving reads), but not to do any deletions. This enables a multi-attached state for tenants,
+  facilitating live migration (this will be articulated in the next RFC that describes HA and
+  migrations). Operating in a stale attachment generation is safe, although for efficiency reasons
+  should not be done over long periods of time, as the data in the stale generation has to avoid
+  doing operations involving deletion, like remote compaction.
+
+### Object Key Changes
+
+#### Generation suffix
+
+All object keys (layer objects and index objects) will contain a numeric suffix that
+advances as the attachment generation and node generation advance.
```

Comment:
Consider if suffix or a prefix would be better:

- S3 LIST operations can use a prefix. Do you ever need to list all files belonging to a particular generation? Or list all files in a particular key-lsn range? Which is more common?

- S3 [partitions data using prefixes](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html). For scalability, it's better to not have hotspot of access with a particular prefix. We are very far having to worry about this in practice, but wanted to mention it.

Reply:
The suffix is intentional, because when we bootstrap a `Timeline` we may need to list indices and pick the one with the newest generation -- added a FAQ item and linked it here.